### PR TITLE
[FLINK-24706][rpc] Forward deserialization errors to returned future

### PR DESCRIPTION
**Pull request by zentol**
Fixes an issue where issues during the deserialization of RPC messages were silently ignored. We now forward the error to the returned future instead.

@tillrohrmann Could you take a look?

### CODE REVIEW

**Comment by zentol** [comment]
@flinkbot run azure

**Comment by zentol** [comment]
hmm...the `TaskManagerProcessFailureBatchRecoveryITCase` is failing pretty reliably now. I'm wondering if this issue was masking another failure as well...

**Comment by zentol** [comment]
That test failure is driving me crazy.

With the original code the test passes.
```
            resultFuture.whenComplete(
                    (resultValue, failure) -> {
                        if (failure != null) {
                            completableFuture.completeExceptionally(
                                    resolveTimeoutException(
                                            failure, callStackCapture, address, rpcInvocation));
                        } else {
                            completableFuture.complete(
                                    deserializeValueIfNeeded(resultValue, method));
                        }
                    });
```
If I move the deserialization into a separate thenApply then it fails.
```
            resultFuture
                    .thenApply(resultValue -> deserializeValueIfNeeded(resultValue, method))
                    .whenComplete(
                            (resultValue, failure) -> {
                                if (failure != null) {
                                    completableFuture.completeExceptionally(
                                            resolveTimeoutException(
                                                    failure,
                                                    callStackCapture,
                                                    address,
                                                    rpcInvocation));
                                } else {
                                    completableFuture.complete(resultValue);
                                }
                            });
```

**Comment by zentol** [comment]
It may be that the test is simply not stable.
What could be happening is that the jobs fails once because the TM loss is noticed by another TM, and then fails again when the JM later attempts to deploy a task to the lost TM.

It's suspicious how _consistently_ the test failed when I switched between the PR and current master, but I also ran into the issue once without any changes...

**Comment by zentol** [comment]
@flinkbot run azure

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaInvocationHandler.java [hunk 1]
```diff
@@ -240,17 +241,20 @@ private Object invokeRpc(Method method, Object[] args) throws Exception {
...
-                        if (failure != null) {
-                            completableFuture.completeExceptionally(
-                                    resolveTimeoutException(
-                                            failure, callStackCapture, address, rpcInvocation));
-                        } else {
-                            completableFuture.complete(
-                                    deserializeValueIfNeeded(resultValue, method));
-                        }
-                    });
+            FutureUtils.forward(
```
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaInvocationHandler.java:244]: I think you are right. We don't need the `forward` here.

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java [hunk 2]
```diff
@@ -67,7 +67,7 @@ public void testTaskManagerFailure(Configuration configuration, final File coord
         ExecutionEnvironment env =
                 ExecutionEnvironment.createRemoteEnvironment("localhost", 1337, configuration);
         env.setParallelism(PARALLELISM);
-        env.setRestartStrategy(RestartStrategies.fixedDelayRestart(1, 1500L));
+        env.setRestartStrategy(RestartStrategies.fixedDelayRestart(10, 1500L));
```
Comment (tillrohrmann) [inline flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java:70]: I don't fully understand why this change is now required. Can we explain why this is the case?

**Review by tillrohrmann — COMMENTED** [review]
Thanks for creating this PR @zentol. I think the change itself looks good. What I don't fully understand is the required change for the `TaskManagerProcessFailureBatchRecoveryITCase`. Can you explain why we need multiple restarts?

**Comment by tillrohrmann** [comment]
From the logs of the failed test run it looks as if we aren't able to detect the dead TM via the lost heartbeats that should be reported to the dead letter queue. From the logs it also looks as if we aren't experiencing a long processing gap which can usually explain a situation like this.

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java (hunk 2 above)
Comment (zentol) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java:70]: The issue depends on when the first heartbeat is sent after the TM went down.

If it happens before the restart, then we properly remove the TM and don't use it later on.
If it does not happen before the restart, then the job fails later on a second time while attempting to deploy task to the same TM.

I think this change just ever so slightly adjusts the timings to make it more common; it can already happen in the current master.

Comment (zentol) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java:70]: - JM sends heartbeat to TM1 (could be transmitted, but may not get a response)
- TM1 is killed
- job restarts (not delayed by missing TM1 because cancelTask RPCs fail immediately)
- JM sends heartbeat to TM1 (could not be transmitted)
- job restarts second time

Comment (tillrohrmann) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java:70]: For this test, we have the heartbeat interval set to 200ms and the restart delay to 1.5s. Hence there should be multiple heartbeats being sent during the restart delay. Moreover, we mark TMs as unreachable after a single lost message. Hence, I can only think of a processing gap on the test machine to explain this situation.

Comment (zentol) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java:70]: I can also reproduce it locally where there are no processing gaps.

Comment (zentol) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java:70]: The restart delay is a fair point, I'll check the logs again.

Comment (tillrohrmann) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java:70]: But maybe the error margin is still too small and we should either increase the number of restart attempts or the restart delay.

For reference I tried to harden the test case via #17107.

Comment (zentol) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java:70]: It looks like the failed heartbeats are being ignored while we are waiting for the restart delay:

```
// 64589-0ef458 is the killed TM

// this is the last heartbeat
17532 o.a.f.r.jobmaster.JobMaster [] - Received heartbeat from 127.0.0.1:64589-0ef458.
...
<start job restart>
<cancel slot requests>
<cleanup partitions>
<various failed cancelTask RPCs>
17740 <reduce resource requirements to 0>
...
17440... JM idling, sending heartbeat requests
19212 o.a.f.r.jobmaster.JobMaster [] - Archive local failure causing attempt 05bcf9159a5a301d2f7b6566111235da to fail
...
19213  o.a.f.r.executiongraph.ExecutionGraph [] - Job Flink Java Job at Mon Nov 01 15:42:30 CET 2021 (4daf5dcbf65f7cd384ac228ad72ab5c6) switched from state RESTARTING to RUNNING.
19777  o.a.f.r.jobmaster.JobMaster [] - TaskManager with id 127.0.0.1:64589-0ef458 is no longer reachable.
19777 o.a.f.r.jobmaster.JobMaster [] - Disconnect TaskExecutor 127.0.0.1:64589-0ef458 because: TaskManager with id 127.0.0.1:64589-0ef458 is no longer reachable.
```

Comment (zentol) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java:70]: I'm wondering if the exception the HeartbeatMonitorImpl sees could be wrapped in a CompletionException...

Comment (zentol) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java:70]: 💢 
```
21440 o.a.f.r.j.JobMaster [] - #handleHeartbeatRpcFailure exception
java.util.concurrent.CompletionException: org.apache.flink.runtime.rpc.exceptions.RecipientUnreachableException
```

Comment (zentol) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java:70]: I suppose stripping the `CompletionException` in the `HeartbeatManagerImpl` should be done in any case because it is so easy to introduce bugs like this.

I'm curious though whether we should revert the `AkkaInvocationHandler` to again to the manual forwarding of the result...

**Comment by zentol** [comment]
How long does it actually take for Akka to consider something to be unreachable?

It takes quite a while for this to show up:
```
441053 [flink-akka.actor.default-dispatcher-23] WARN  akka.remote.transport.netty.NettyTransport [] - Remote connection to [null] failed with java.net.ConnectException: Connection refused: no further information: /127.0.0.1:58498
441056 [flink-akka.actor.default-dispatcher-23] WARN  akka.remote.ReliableDeliverySupervisor [] - Association with remote system [akka.tcp://flink@127.0.0.1:58498] has failed, address is now gated for [50] ms. Reason: [Association failed with [akka.tcp://flink@127.0.0.1:58498]] Caused by: [java.net.ConnectException: Connection refused: no further information: /127.0.0.1:58498]
```

to show up, and after that we immediately throw `RecipientUnreachableExceptions.`

**Comment by zentol** [comment]
So yeah, increasing the restart delay to ~5 seconds would probably do the trick.

File: flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerProcessFailureBatchRecoveryITCase.java]: Hmm, with removing the forwarding we do indeed change the exception type for consumers of the returned future. If we don't want to accept this risk, then I think it is probably safer to revert this change.

In general, consumers of futures should handle the `CompletionException` case though (e.g. via stripping or using `ExceptionUtils.containsCause`).

**Comment by tillrohrmann** [comment]
I don't know exactly how fast Akka can detect it. I would assume that this can happen quite quickly once the TCP connection is terminated.

**Comment by zentol** [comment]
> This has never been a problem though.

Isn't that because previously we still waited for the heartbeat to time out (10 seconds in this test)?

**Comment by tillrohrmann** [comment]
Maybe. You could enable debug logs and check whether we send heartbeat requests after the TM died. If these requests are not failed, then it is most likely because Akka hasn't detected that the connection is dead yet.

**Comment by tillrohrmann** [comment]
The advantage of increasing the restart attempts instead of the delay is that the test will on average run faster. With the increased delay, the execution of the test will take at least as much longer as you've increased the delay.

**Comment by tillrohrmann** [comment]
> I don't have a strong opinion on this. A single restart results in cleaner logs.
If we're aiming to do at most 1-2 restarts then the difference will be negligible either way (+-2s)🤷 The test already needs 30-50 seconds.

Fair enough, then lets go with a single restart and an extended delay.
